src/main/java/io/appium/java_client/proxy/ProxyListenersContainer.java

```diff
@@ -39,25 +37,17 @@ public static synchronized ProxyListenersContainer getInstance() {
     // we had to change it to a list, which has O(N). The reason for that is that
     // maps implicitly call `hashCode` API on instances, which might not always
     // work as expected for arbitrary proxies
-    private final List<Pair<WeakReference<?>, Collection<MethodCallListener>>> listeners = new LinkedList<>();
-
-    private static class Pair<K, V> {
-        private final K key;
-        private final V value;
-
-        public Pair(K key, V value) {
-            this.key = key;
-            this.value = value;
-        }
-
-        public K getKey() {
-            return key;
-        }
-
-        public V getValue() {
-            return value;
-        }
-    }
+    // As of (this version?) we are now using a LinkedHashMap as a caching mechanism. This can be overridden by a System
+    // Property value "appium.override.listener.limit": default 5000.
+    private final LinkedHashMap<Object, HashSet<MethodCallListener>>
```

Comment:
I am not sure we want to have LRU-like structure here. We don't want older listeners to be dropped based on the amount of newer ones. We only want them to be dropped if the amount of references to them is zero (e.g. ready for garbage collect)